### PR TITLE
Init commit

### PR DESCRIPTION
**Pull request by ScorpioCPH**
Hi, this PR implements the following features:

- Register a new custom resource of type `ResourceClass` using a CustomResourceDefinition (CRD), support both namespace and cluter scopes.
- Create/Delete/Get/List/Update instances of this new resource type `ResourceClass`.
- Attach labels on the nodes for this `ResourceClass`.
- Mark taints on nodes to declare these nodes are dedicated if needed.

Future works:

- Handle `matchExpressions` for this `ResourceClass`.
- Bind `ResourceClass` with `ResourceName` and `Devices` which are advertised by device plugin.

@ddysher @NickrenREN PTAL.

### CODE REVIEW

**Comment by ddysher** [comment]
@ScorpioCPH Can u send proposal first and link the resource class proposal from upstream?

We can find folks in dev team to help review the PR if docs are ready.

**Comment by ScorpioCPH** [comment]
@ddysher Sure.

- upstream [proposal](https://github.com/kubernetes/community/pull/782)
- oscenter [proposal](https://github.com/caicloud/oscenter/pull/23)

**Comment by ScorpioCPH** [comment]
@gaocegege @DjangoPeng @zjx-caicloud comments are welcome :)

**Comment by gaocegege** [comment]
Thanks for your information and it works for me, although I do not know the meaning of `devicesID` in examples :thinking:

**Comment by ScorpioCPH** [comment]
@gaocegege It is exposed by device plugin, here is [design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md), FYI.

**Comment by ddysher** [comment]
@ScorpioCPH  let's make it an umbrella issue and break it into different PRs. This PR is impossible to review

**Comment by ddysher** [comment]
Suggestion for breaking down:

- API and auto-generated code
- README: not just how to build the repository, but also links to design docs, dev status, etc
- Main logic and Makefile
- Example

File: README.md [hunk 1]
```diff
@@ -1 +1,24 @@
...
+This repository implements a `resource` controller for watching `ResourceClass` resources as
+defined with a `CustomResourceDefinition` (CRD).
+
+This is a part of POC implementation of Resource Class with `CRD`, `Custom Scheduler` and `Device Plugin`.
+
+## Build & Run
+
+```sh
+# build
+$ make resource-controller
```
Comment (ddysher) [inline README.md]: It won't build in my environment (I have latest kubernetes code).  Please add vendor in the PR.

File: pkg/apis/resourceclass/v1alpha1/doc.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v1alpha1 is the v1alpha1 version of the API.
+// +k8s:deepcopy-gen=package,register
+// +groupName=resourceclass.controller.k8s.io
```
Comment (ddysher) [inline pkg/apis/resourceclass/v1alpha1/doc.go]: what does `controller` stands for here?

File: pkg/apis/resourceclass/v1alpha1/register.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2015 The Caicloud Authors.
```
Comment (ddysher) [inline pkg/apis/resourceclass/v1alpha1/register.go]: 2015?

File: pkg/apis/resourceclass/v1alpha1/types.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Required. Spec defines resources
+	Spec ResourceClassSpec `json:"spec"`
+	// Required.
+	Status ResourceClassStatus `json:"status"`
+}
+
+// ResourceClassSpec is the spec for a ResourceClass resource
+type ResourceClassSpec struct {
+	// Required. ResourceName is advertised by device plugin.
```
Comment (ddysher) [inline pkg/apis/resourceclass/v1alpha1/types.go]: Why do you add `ResourceName` here?  Can I have a resource class with multiple resources advertised from device plugin?

File: pkg/apis/resourceclass/v1alpha1/types.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	ResourceName string `json:"resourceName"`
+	// Required. DevicesIDs is advertised by device plugin, we cached here.
+	DevicesIDs []string `json:"deviceIDs"`
+	// Required. A list of resource selector requirements.
+	// The requirements are ANDed.
+	MatchExpressions []ResourceSelectorRequirement
+	// Required. An array of available nodes for this resource class.
+	// If empty then this resource can't be scheduled.
+	Nodes []NodeSpec `json:"nodes"`
```
Comment (ddysher) [inline pkg/apis/resourceclass/v1alpha1/types.go]: `Node` come and go, do we need to cache it here? 

 (Even if the answer is yes, we want to put it under status)

**Pull request closed without merging**